lib/logstash/filters/json.rb

```diff
@@ -46,54 +46,87 @@ class LogStash::Filters::Json < LogStash::Filters::Base
+        return
       end
 
-      # If no target, we target the root of the event object. This can allow
-      # you to overwrite @timestamp and this will typically happen for json
-      # LogStash Event deserialized here.
-      if !@target && event.timestamp.is_a?(String)
-        event.timestamp = LogStash::Timestamp.parse_iso8601(event.timestamp)
+      # TODO: (colin) the timestamp initialization should be DRY'ed but exposing the similar code
+      # in the Event#init_timestamp method. See https://github.com/elastic/logstash/issues/4293
+
+      parsed_timestamp = parsed.delete(LogStash::Event::TIMESTAMP)
+      begin
+        timestamp = parsed_timestamp ? LogStash::Timestamp.coerce(parsed_timestamp) : nil
+      rescue LogStash::TimestampParserError => e
+        timestamp = nil
       end
 
-      filter_matched(event)
-    rescue => e
-      tag = "_jsonparsefailure"
-      event["tags"] ||= []
-      event["tags"] << tag unless event["tags"].include?(tag)
-      @logger.warn("Trouble parsing json", :source => @source,
-                   :raw => event[@source], :exception => e)
-      return
+      inject_fields(event, parsed)
+
+      if parsed_timestamp
+        if timestamp
+          event.timestamp = timestamp
+        else
+          event.timestamp = LogStash::Timestamp.new
+          @logger.warn("Unrecognized #{LogStash::Event::TIMESTAMP} value, setting current time to #{LogStash::Event::TIMESTAMP}, original in #{LogStash::Event::TIMESTAMP_FAILURE_FIELD} field", :value => parsed_timestamp.inspect)
+          event.tag(LogStash::Event::TIMESTAMP_FAILURE_TAG)
+          event[LogStash::Event::TIMESTAMP_FAILURE_FIELD] = parsed_timestamp.to_s
+        end
+      end
+      # event.to_hash.merge! parsed
     end
 
-    @logger.debug("Event after json filter", :event => event)
+    filter_matched(event)
+
+    @logger.debug? && @logger.debug("Event after json filter", :event => event)
+  end
 
-  end # def filter
+  private
 
-end # class LogStash::Filters::Json
+  # TODO: (colin) this will need to be DRY'ed with filter-kv once https://github.com/elastic/logstash/issues/4293 is done.
+  #
+  # recursiverly inject the data Hash content into event. the data parameter must be Hash on initial call.
+  #
+  # this is important to do here because we cannot assume in the future the in-place mutability of the Event#to_hash structure.
+  # TODO: (colin) we can certainly optimize this by introducing a new Event API to directly set fields without going through the
+  # field reference API Event#[]=
+  #
+  # @param event [Event] the event to inject fields in
+  # @param path [String] the field reference to inject the data value
+  # @param data [Hash|Array|String] the data to inject in the given path, when data is a Hash inject_fields will be
+  #   recursively called for each key/values of data
+  def inject_fields(event, data, path = nil)
```

Comment:
same comment as in https://github.com/logstash-plugins/logstash-filter-kv/pull/21#discussion_r46525441
maybe I'm missing something obvious here :)